beta/src/pages/learn/adjusting-effect-dependencies.md

```diff
@@ -0,0 +1,1345 @@
+    port: port,
+    auth: {
+      token: token
+    }
+  };
+
+  useEffect(() => {
+    const connection = createConnection(roomId, options);
+    connection.connect();
+    return () => connection.disconnect();
+  }, [roomId, options]); // ✅ All dependencies declared
+  // ...
+}
+```
+
+This code has the same issue as [the first example.](#removing-object-dependencies) When the `message` changes, the `options` object gets re-created. Since it's different on every render, the Effect will re-run and reconnect to the server on every keystroke.
+
+**In this example, the correct solution is to move the object *inside* the Effect itself:**
+
+```js {7-12,16}
+function ChatRoom({ roomId }) {
+  const [message, setMessage] = useState('');
+  const { port } = useContext(SettingsContext);
+  const { token } = useContext(AuthContext);
+
+  useEffect(() => {
+    const options = { // ✅ Not a dependency
+      port: port,
+      auth: {
+        token: token
+      }
+    };
+    const connection = createConnection(roomId, options);
+    connection.connect();
+    return () => connection.disconnect();
+  }, [roomId, port, token]); // ✅ All dependencies declared
+  // ...
+}
+```
+
+Previously, `options` needed to be a dependency because it was declared directly in the component body. Now, your Effect *creates* the `options` object based on the `port` and `token` from the component body. This is why `port` and `token` must be dependencies now. Both `port` and `token` are [reactive values](#every-reactive-value-becomes-a-dependency). They could change over time, and then you would *want* the Effect to re-run. This is why declaring them as dependencies makes sense.
+
+Since the `options` object is no longer a dependency, it doesn't make the Effect re-run unnecessarily.
+
+#### Solution 3: Recreate the object from primitives {/*solution-3-recreate-the-object-from-primitives*/}
+
+In the above example, both `port` and `token` are [primitive data types](https://developer.mozilla.org/en-US/docs/Glossary/Primitive). With primitive data types likes strings, numbers, and booleans, you don't need to worry about whether the value is *actually* different or accidentally different (like with objects). **Working with Effects is easier when you try to keep their dependencies primitive.**
+
+Let's say you receive the whole `options` object as a prop:
+
+```js {1,8}
+function ChatRoom({ roomId, options }) { // Unclear: Could be a new object every time
+  const [message, setMessage] = useState('');
+
+  useEffect(() => {
+    const connection = createConnection(roomId, options);
+    connection.connect();
+    return () => connection.disconnect();
+  }, [roomId, options]); // ✅ All dependencies declared
+  // ...
+}
+```
+
+You can't know whether the parent component will pass an `options` object that's "stable" (i.e. does not change between re-renders) or if it's going to create a completely different object every time that it re-renders.
```

Comment:
this example rewrite is fine but I feel like it's also legit to have stable identity as part of the contract for the component, especially if the options type is from a third-party (here, websocket) library – saying "connectionOptions prop; each time it changes the connection will be reestablished; please useMemo" – wdyt?